### PR TITLE
Continue

### PR DESCRIPTION
**Pull request by nickvergessen**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -37,21 +39,20 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
 	switch {
 	case err == sql.ErrNoRows:
 		// Device not found
-		insert, error := db.Prepare("INSERT INTO `devices` (`user_public_key`, `device_identifier`, `push_token`, `push_token_hash`) VALUES (?,?,?,?)")
-		if error != nil {
+		insert, err := db.Prepare("INSERT INTO `devices` (`user_public_key`, `device_identifier`, `push_token`, `push_token_hash`) VALUES (?,?,?,?)")
```
Comment (nickvergessen) [inline main.go]: > declaration of err shadows declaration at main.go:22:  (vetshadow)at line 42 col 

@leonklingele any idea about this message from go-meta-linter?

File: main.go [hunk 2]
```diff
@@ -17,6 +17,14 @@ import (
...
+	errInvalidUserPublicKey         = errors.New("Invalid user public key supplied")
+	errInvalidDeviceIdentifier      = errors.New("Invalid device identifier supplied")
+	errInvalidUserPublicKeyMismatch = errors.New("User public key does not belong to device identifier")
+	errInvalidSignature             = errors.New("Invalid signature supplied")
+	errInvalidSignatureMismatch     = errors.New("Invalid signature supplied")
+)
+
 func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
 
 	db, err := sql.Open("mysql",
```
Comment (nickvergessen) [inline main.go]: I thought the same, wanted to ask you tomorrow

File: main.go [hunk 3]
```diff
@@ -26,9 +34,11 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
...
-	var dbUserPublicKey, dbDeviceIdentifier, dbPushToken, dbPushTokenHash string
-	row := db.QueryRow("SELECT * FROM `devices` WHERE `device_identifier` = ?", deviceIdentifier)
-	err = row.Scan(&dbUserPublicKey, &dbDeviceIdentifier, &dbPushToken, &dbPushTokenHash)
+	defer db.Close()
+
+	var dbUserPublicKey string
+	row := db.QueryRow("SELECT `user_public_key` FROM `devices` WHERE `device_identifier` = ?", deviceIdentifier)
+	err = row.Scan(&dbUserPublicKey)
 
 	h := sha512.New()
```
Comment (nickvergessen) [inline main.go]: The problem is that we need the original value (to push) and can only search on the hashed value (nextcloud instance only has this)

File: main.go [hunk 4]
```diff
@@ -37,44 +47,126 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
 	switch {
```
Comment (nickvergessen) [inline main.go]: copied from sample :-X

File: main.go [hunk 5]
```diff
@@ -37,44 +47,126 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
...
 
-		insert.Exec(
-			userPublicKey,
-			deviceIdentifier,
-			pushToken,
-			pushTokenHash,
-		)
-		log.Println("DEVICE_INSERTED")
+		if _, err = insert.Exec(userPublicKey, deviceIdentifier, pushToken, pushTokenHash); err != nil {
+			log.Println(err)
```
Comment (nickvergessen) [inline main.go]: Logging is currently for debugging only
Real logs will be put in place before going productive. They need more thinking on what we want to log etc.

File: main.go [hunk 6]
```diff
@@ -37,44 +47,126 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
...
+		log.Println("Device inserted")
 
 	case err != nil:
-		log.Fatal(err)
+		log.Println(err)
 		return err
 
 	default:
 		// Device found: update if user key is the same
 		if userPublicKey != dbUserPublicKey {
```
Comment (nickvergessen) [inline main.go]: Doesn't matter, the signature verification is what matters more, if you end up in here you are not hacking but your key changed.

File: main.go [hunk 7]
```diff
@@ -37,44 +47,126 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
...
 
+	var dbUserPublicKey string
+	row := db.QueryRow("SELECT `user_public_key` FROM `devices` WHERE `device_identifier` = ?", deviceIdentifier)
+	err = row.Scan(&dbUserPublicKey)
+
+	switch {
+	case err == sql.ErrNoRows:
+		// Device not found
+		log.Println("To be deleted device could not be found")
+		return nil
```
Comment (nickvergessen) [inline main.go]: Not really, it was successful, the device identifier is now unknown to the server

File: main.go [hunk 8]
```diff
@@ -37,44 +47,126 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
...
+	}
+
+	return nil
+}
+
+func validateDeviceIdentifier(deviceIdentifier, deviceIdentifierSignature, userPublicKey string) error {
+	if deviceIdentifier == "" || len(deviceIdentifier) != 128 {
+		return errInvalidDeviceIdentifier
+	}
+	if deviceIdentifierSignature == "" || len(deviceIdentifierSignature) < 256 || len(deviceIdentifierSignature) > 512 {
```
Comment (nickvergessen) [inline main.go]: simplification works, the check looks incorrect, but we want to catch outside values, so its right ;)

File: main.go (hunk 6 above)
Comment (leonklingele) [inline main.go]: > if you end up in here you are not hacking but your key changed.

Not true. This check here is the __most important__ check of the entire system.

The signature check only ensures that _some_ user-provided public key can be used to verify that _some_ user-provided device identifier was signed by the corresponding private key to produce _some_ user-provided signature. Everyone can pass this check for any device identifier:
Say you want to pass the check for id = "test":
1. Create a new public / private key pair
2. Sign the id "test" with your private key to get a signature sig
3. Submit sig, id and your public key to this service
4. Voilà. Validation will succeed

__EDIT__:
The more I think about it, the more this system looks utterly complicated to me. Why are you even using public key crypto here? A password-based authentication (bcrypt, scrypt, pbkdf2, whatever) between a device id's owner and this service should be enough, and much faster. No need to use RSA here?

**Comment by leonklingele** [comment]
Looks like I was a bit over-excited of requiring everything to run in constant time, but in this case it really isn't needed. :'(
Still, what advantage does public key crypto bring in?

File: main.go (hunk 6 above)
Comment (leonklingele) [inline main.go]: Sorry for reposting this comment, but Github didn't use it for this line-conversation when I replied via Email:

Looks like I was a bit over-excited of requiring everything to run in constant time, but in this case it really isn't _needed_. :'(
Still, what advantage does public key crypto bring in?

File: main.go [hunk 9]
```diff
@@ -37,44 +47,126 @@ func savePushToken(pushToken, deviceIdentifier, userPublicKey string) error {
...
+	if err != nil {
+		return errInvalidSignature
+	}
+
+	if userPublicKey == "" || len(userPublicKey) != 450 {
+		return errInvalidUserPublicKey
+	}
+
+	block, _ := pem.Decode([]byte(userPublicKey))
+	if block == nil {
```
Comment (nickvergessen) [inline main.go]: Then it's not defined anymore for `x509.ParsePKIXPublicKey` later on?

**Comment by nickvergessen** [comment]
I'll merge this to continue in a new PR

**Pull request merged**